pennylane/resource/estimation.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for resource estimation.
+"""
+
+from collections import defaultdict
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.operation import AnyWires, Operation
+
+
+class Resource:
+    r"""Create a resource object for storing quantum resource information."""
+
+    def __init__(self):
+        self.num_wires = 0
+        self.num_gates = 0
+        self.gate_types = defaultdict(int)
+        self.depth = 0
+        self.shots = 0
```

Comment:
One worry I have is how to deal with custom resources, e.g., if a user wants to define an additional `my_resource` on their custom operation. How are `my_resource` resources combined in a circuit? What do we do if a circuit consists of some custom and some standard operations (which don't have the `my_resource` resource attached) - would users need to specify `my_resource` even for operations like `Hadamard`?

And how important do we expect it to be for users to define a custom resource? I'm leaning towards it not being too relevant, but curious to hear your thoughts?

Reply:
> if a user wants to define an additional my_resource on their custom operation. How are my_resource resources combined in a circuit?

It is pretty challenging for the estimator to deal with this type of resource types. We probably have to fix the resource types and not allow the users to add new ones to their custom operations. 

> What do we do if a circuit consists of some custom and some standard operations (which don't have the my_resource resource attached)
For standard operators, we estimate the gate recourse based on the decomposition provided in the operation gates. 

> would users need to specify my_resource even for operations like Hadamard?
They can define a `CustomOperation` that inherits from the standard operation and add  `resources` method to it with their own specification of the resource types. e.g., if they have new decomposition of the `qml.DoubleExcitation`, they can do:

```
class CustomOperation(qml.DoubleExcitation):
    num_wires = 4

    def resources(self):
        resource = Resource()
        resource.num_gates = 11
        resource.gate_types = {"T": 5, "CNOT": 6}
        resource.depth = 11
        return resource
```